Source/Details/ASThread.h

```diff
@@ -31,6 +29,60 @@ static inline BOOL ASDisplayNodeThreadIsMain()
   return 0 != pthread_main_np();
 }
 
+/**
+ * Adds the lock to the current scope.
+ *
+ * A C version of the C++ lockers. Pass in any id<NSLocking>.
+ * One benefit this has over C++ lockers is that the lock is retained. We
+ * had bugs in the past where an object would be deallocated while someone
```

Comment:
I recall there was a solution to this issue that prevented deallocation when the lock was owned. Is this saying the problem was not fully solved, even in ASDisplayNode?

My main concern with this patch is of course not the message sends themselves, but the retain / release traffic if we were to use these facilities in core codepaths would be quite significant - and almost always unnecessary. That is to say, only one reference to an object is needed for it to be safe, and in most codepaths there does exist such a reference.

Since we're mainly changing the subclasses instead of ASDisplayNode, that should be OK. That said, the optimizations in ASNetworkImageNode were developed through profile-driven improvements. It may be appropriate to use the Unowned variant in core code such as ASNetworkImageNode where there is a known reference, e.g. if we have a keepalive from our view object which is still in a hierarchy then there is really no reason to retain here (and we could instead lock the object before clearing that keepalive).